Stock ownership availability rules

lepistone opened the pull request.

lepistone committed:
stock_ownership_availability_rules stub

lepistone committed:
fist passing test on quant reservation

lepistone committed:
full, partial reservation with no owner

lepistone committed:
it does not reserve other people's quants

lepistone committed:
handle move with owner

lepistone committed:
add tests for when the quant is owned by us

For our purposes, no owner behaves the same way as company owner.

lepistone committed:
handle delivery with company owner

lepistone committed:
make the owner of the quant required

This requires a pre init hook for installation and a create().

lepistone committed:
automatically enable group to show owners

lepistone committed:
open quants of a product grouped by owner

pedrobaeza commented on stock_ownership_availability_rules/README.rst:
```diff
@@ -0,0 +1,7 @@
+Stock Ownership Availability Rules
+==================================
+
```
Can you please explain the purpose of the module?

lepistone replied:
done!

lepistone committed:
test: quants have a default owner

lepistone committed:
set quant owner to company on reception

lepistone committed:
remove mutable default parameter

lepistone committed:
add doc

lepistone committed:
remove tests with quants with no owner

These tests are not necessary anymore, because the owner of the quant is
now required. A separate test to check that default is already in place.

Even before, "no owner" and "owned by myself" meant the same thing.

The tests were green.

gurneyalex commented on stock_ownership_availability_rules/__openerp__.py:
```diff
@@ -0,0 +1,34 @@
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+{'name': 'Stock Ownership Availability Rules',
+ 'summary': 'XXX',
```
Can you remove the entry or provide a meaningful summary?

gurneyalex commented on stock_ownership_availability_rules/model/product.py:
```diff
@@ -0,0 +1,14 @@
+        result['context'] = (
+            "{'search_default_locationgroup': 1, "
+            "'search_default_ownergroup': 1, "
+            "'search_default_internal_loc': 1}"
```
curious... You can't use a real dict here?!

lepistone replied:
I think I can indeed. The string comes from the overridden method in addons/stock/product.py.

gurneyalex commented on stock_ownership_availability_rules/model/quant.py:
```diff
@@ -0,0 +1,59 @@
+            prefered_domain_list = []
+
+        my_partner = location.company_id.partner_id
+        if restrict_partner_id == my_partner.id or not restrict_partner_id:
```
Several points here:
1. i find the identifier 'my_partner' confusing in this context. 
2. nothing seems to be done about location.partner_id.id? Is it normal?
3. the condition in the test is not clear to me. I would have expected `and restrict_partner_id` instead of `or not restrict_partner_id`. Am I missing something obvious?

lepistone replied:
1. it means "my company's partner". What do you suggest?
2. No, I agree with you. If it is set, we can use that for "my_partner" as a first choice.
3. it means in english: "If there is no restriction, or if the restriction is on my_partner" which mean the same.

In general, I'm open to suggestions to refactor this.

gurneyalex replied:
re 1. -> location_partner_id

lepistone replied:
Refactored and docstring added.

gurneyalex replied:
re 3: -> OK got it.  I suggest you write the condition in the same way you've explained it in english (i.e. permute the operands of `or` which sounds more natural to me). Side question: is it possible for `my_partner.id` to be `False`?

lepistone committed:
add module summary

lepistone committed:
imp docstring

lepistone committed:
refactor and doc the computation of the domain

I wrote this code when the owner of quants was not required, so I needed
to handle more cases. Now the logic can be simplified.

Moreover, If a location has a partner_id, I use that as a definition of
"own" stock.

lepistone commented:
@gurneyalex you're right with your "side question": if a location has no partner, and no company (neither is required), what can I use as "own" stock? `self.env.user.company_id.partner_id`?

gurneyalex commented:
@lepistone: sounds good to me.

gurneyalex commented:
:+1:

yvaucher commented on stock_ownership_availability_rules/model/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
```
Missing license header

yvaucher commented on stock_ownership_availability_rules/model/product.py:
```diff
@@ -0,0 +1,14 @@
+from openerp import models
```
Missing header

yvaucher commented on stock_ownership_availability_rules/model/quant.py:
```diff
@@ -0,0 +1,64 @@
+from openerp import models, api, fields
```
No header

yvaucher commented on stock_ownership_availability_rules/model/quant.py:
```diff
@@ -0,0 +1,64 @@
+        if domain is None:
+            domain = []
+        if prefered_domain_list is None:
+            prefered_domain_list = []
```
~~Is this check necessary ? Can't you just throw prefered_domain_list as it is to super ?~~

Ok `prefered_domain_list=[]` in signature in stock.py :-/

yvaucher commented:
Apart from missing license header LGTM :+1:

lepistone committed:
added missing copyright headers

lepistone commented:
@yvaucher I added the missing headers. Feel free to merge :smile_cat:

yvaucher commented:
Thanks
